iwyu_output.cc

```diff
@@ -1889,6 +1893,41 @@ size_t PrintableDiffs(const string& filename,
 
 }  // namespace internal
 
+void IwyuFileInfo::HandlePreprocessingDone() {
+  // Check macros defined by includer.  Requires file preprocessing to be
+  // finished to know all direct includes and all macro usages.
+  bool should_report_violations = ShouldReportIWYUViolationsFor(file_);
+  std::list<const FileEntry*> direct_macro_use_includees;
+  std::set_intersection(macro_users_.begin(), macro_users_.end(),
+                        direct_includes_as_fileentries_.begin(),
+                        direct_includes_as_fileentries_.end(),
+                        std::inserter(direct_macro_use_includees,
+                                      direct_macro_use_includees.end()));
+  for (const FileEntry* macro_use_includee : direct_macro_use_includees) {
+    if (should_report_violations) {
+      ERRSYM(file_) << "Keep #include " << macro_use_includee->getName()
```

Comment:
You are correct, 'control' macros can cause keeping unnecessary includes. Some cases are somewhat alleviated by putting these macros and includes in non-associated header like `config.h`.

One way to improve the situation is to distinguish 'control' macros, macros that occur only in #if, #ifdef, etc. This way we'll be able to tell if content of included file is required to be used to keep it. Because for x-macros we still want to keep #include even if it looks like it is unused.